src/SofaCaribou/Mass/CaribouMass.inl

```diff
@@ -0,0 +1,506 @@
+        gauss_nodes.resize(element.number_of_gauss_nodes());
+    }
+
+    const auto nb_of_gauss_nodes = gauss_nodes.size();
+    for (std::size_t gauss_node_id = 0; gauss_node_id < nb_of_gauss_nodes; ++gauss_node_id) {
+        const auto & g = element.gauss_node(gauss_node_id);
+
+        const auto J = element.jacobian(g.position);
+        const auto detJ = std::abs(J.determinant());
+
+        // Derivatives of the shape functions at the gauss node with respect to global coordinates x,y and z
+        const Vector<NumberOfNodesPerElement> N = element.L(g.position);
+
+
+        GaussNode & gauss_node = gauss_nodes[gauss_node_id];
+        gauss_node.weight               = g.weight;
+        gauss_node.jacobian_determinant = detJ;
+        gauss_node.N                    = N;
+    }
+
+    return gauss_nodes;
+}
+
+template <typename Element>
+void CaribouMass<Element>::initialize_elements()
+{
+    using namespace sofa::core::objectmodel;
+
+    sofa::helper::AdvancedTimer::stepBegin("CaribouMass::initialize_elements");
+
+    if (!this->mstate)
+        return;
+
+    // Resize the container of elements'quadrature nodes
+    const auto nb_elements = this->number_of_elements();
+    if (p_elements_quadrature_nodes.size() != nb_elements) {
+        p_elements_quadrature_nodes.resize(nb_elements);
+    }
+
+    // Translate the Sofa's mechanical state vector to Eigen vector type
+    sofa::helper::ReadAccessor<Data<VecCoord>> sofa_x0 = this->mstate->readRestPositions();
+    const Eigen::Map<const Eigen::Matrix<Real, Eigen::Dynamic, Dimension, Eigen::RowMajor>>    X0      (sofa_x0.ref().data()->data(), sofa_x0.size(), Dimension);
+
+    // Loop on each element and compute the shape functions and their derivatives for every of their integration points
+    for (std::size_t element_id = 0; element_id < nb_elements; ++element_id) {
+
+        // Get an Element instance from the Domain
+        const auto initial_element = this->topology()->element(element_id);
+
+        // Fill in the Gauss integration nodes for this element
+        p_elements_quadrature_nodes[element_id] = get_gauss_nodes(element_id, initial_element);
+    }
+
+    // Compute the volume
+    Real v = 0.;
+    for (std::size_t element_id = 0; element_id < nb_elements; ++element_id) {
+        for (const auto & gauss_node : gauss_nodes_of(element_id)) {
+            // Jacobian of the gauss node's transformation mapping from the elementary space to the world space
+            const auto detJ = gauss_node.jacobian_determinant;
+
+            // Gauss quadrature node weight
+            const auto w = gauss_node.weight;
+
+            v += detJ*w;
+        }
+    }
+
+    msg_info() << "Total mass of the geometry is " << v*d_density.getValue();
+
+    sofa::helper::AdvancedTimer::stepEnd("CaribouMass::initialize_elements");
+}
+
+template<typename Element>
+void CaribouMass<Element>::addForce(const sofa::core::MechanicalParams * /*mparams*/, CaribouMass::DataVecDeriv & d_f,
```

Comment:
Yeah it should in my opinion, but since I want people to be able to use this component almost the same way as SOFA's one, but with non-linear element compatibility, I figured I would keep it that way for now.